### ISSUE TITLE
FR: Enable blackbricks to use most of sqlparse options

### ISSUE THREAD

**Issue by de-hru** [inspera/blackbricks#31]
Relates to:
- Metadata-Version: 2.1
- Name: blackbricks
- Version: 1.0.1

<br><br>
**AS-IS:** (in blackbricks.py)
```
sqlparse.format(
                "\n".join(sql_lines), reindent=True, keyword_case=sql_keyword_case
```

- **reindent**
  - is set to True always when sqlparse is called by blackbricks
- **keyword_case**
  - defaults to keyword_case=upper, and can be influenced using --no-sql-upper … in this case keyword_case=lower will be used

<br><br>
**TO-BE:** (requested feature)
- Improve blackbricks to be able to set/use **(nearly) all** sqlparse options. For this, please:
  - implement additional sqlparse-options which can be used in blackbricks
  - pass them to sqlparse
  - add default value(s) for these sqlparse options (similar to: **--sql-upper** is default (upper(KEYWORDS), but one can optionally overwrite this using **--no-sql-upper** (lower(KEYWORDS)) in order to ensure downwards comaptibility
    - so ... nobody using blackricks already would need to change anything, because blackbricks would set DEFAULT sqlparse options by default!
    - BTW, I've already checked sqlparse for DEFAULT values ... and I've put them into the list beneath as _<DEFAULT_VALUE> is DEFAULT in sqlparse_

- Here's the list of desired/commonly-used options when formatting SQL source code:
  - **--sql-identifier-case**
     - Allow: upper, lower, capitalize
     - default = n/a
       - _None is DEFAULT in sqlparse?!?_
  - **--sql-strip_comments**
     - Allow: True, False
     - default = False
       - _False is DEFAULT in sqlparse_
  - **--sql-reindent**
     - Allow: True, False
     - default = False
       - _False is DEFAULT in sqlparse_
  - **--sql-reindent_aligned**
     - Allow: True, False
     - default = False
       - _False is DEFAULT in sqlparse_
  - **--sql-use_space_around_operators**
     - Allow: True, False
     - default = False
       - _False is DEFAULT in sqlparse_
  - **--sql-indent_tabs**
    - Allow: True, False
    - default = False
      - _False is DEFAULT in sqlparse_
  - **--sql-indent_width**
    - Allow: smallint or 1..8 or 1..256
    - default = 2
      - _2 is DEFAULT in sqlparse_
  - **--sql-wrap_after**
    - Allow: smallint or 1..8 or 1..256
    - default = 0
      - _0 is DEFAULT in sqlparse_
  - **--sql-comma_first**
    - Allow: True, False
    - default = False
      - _False is DEFAULT in sqlparse_

Dejan Hrubenja dejan.hrubenja@mercedes-benz.com Mercedes-Benz Tech Innovation GmbH 
https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md

**Comment by de-hru** [comment]
@bsamseth Would you appreciate it, if I would implement my FR in a fork and add a PR?

**Comment by bsamseth** [comment]
This is certainly a possibility, and if it was to be added, I think you've outlined a good approach that ensures the defaults retain the existing functionality.

I am, however, a bit hesitant to add all these options. I think blackbricks ought to keep the spirit of black, which purposefully remains mostly non-configurable. There is a lot of value in taking away even the option to consider how formatting is done, and just have the tools do it for you, and I think this is the absolute best part of black.

Of course, SQL is not Python, and to my knowledge there is not a comparable de-facto standard way to format SQL. But for this project specifically, I think a good reference would be to try to replicate the formatting you get when you use the "Format SQL" option in the Databricks notebook UI. Since blackbricks is specific to Databricks, using them as the canonical reference for how to format SQL could be a good idea. The benefit is that you could use the built in formatter while developing without causing large reformats.

TLDR; My suggestion is to _not_ add any additional formatting options, but rather to change the default format options such that we replicate the formatting produced by the built in "Format SQL" option in Databricks. It remains to be seen if this is possible to do perfectly with sqlparse, but I'm hopeful. We should keep the UPPERCASE/lowercase switch.

Is that something you'd be willing to take a stab at? If so, that would be great! :)

**Comment by de-hru** [comment]
> [...] but rather to change the default format options such that we replicate the formatting produced by the built in "Format SQL" option in Databricks. It remains to be seen if this is possible to do perfectly with sqlparse, but I'm hopeful. 

**Some ugly SQL code:**

![image](https://user-images.githubusercontent.com/120418763/215490901-edc6c18d-65ad-489d-8b00-ca02cc83fd21.png)

<br><br>**Result with the built in "Format SQL" option in Databricks**
**I don't like the way how Databricks UI formats SQL code. Most of my colleagues don't like it too.**

![image](https://user-images.githubusercontent.com/120418763/215491552-29de98d9-4c5b-4059-94a6-79df75b5543c.png)

<br><br>**pay attention to keyword formatting with the built in "Format SQL" option in Databricks**

![image](https://user-images.githubusercontent.com/120418763/215491130-d319b2bb-25c9-4ba7-ac12-d25382280b83.png)

<br><br>**Result with the blackbricks/sqlparse - current options**

![image](https://user-images.githubusercontent.com/120418763/215491307-0b039fdc-7d23-4b69-ab60-827eb8b6db40.png)

**Comment by de-hru** [comment]
> [...] there is not a comparable de-facto standard way to format SQL.

I totally agree with you.

> [...] to try to replicate the formatting you get when you use the "Format SQL" option in the Databricks notebook UI. Since blackbricks is specific to Databricks, using them as the canonical reference for how to format SQL could be a good idea. The benefit is that you could use the built in formatter while developing without causing large reformats.

As far as I can see, we cannot see which formatting rules Databricks UI uses. And they might change it from time to time.

Normally, we never use the built-in "Format SQL" function in the Databricks UI. We use some "external" formatting tools after we have completed the implementation:
- copy SQL code from Databricks & paste it into an "external" formatter
- re-format there
- copy&paste back to Databricks

IMHO more people/projects would use blackbricks if SQL formatting options would be a bit more influenceable.
Right now we can not influence:
- indent_type ==> TABs or SPACEs in indentation (indent_tabs [False|True]
- indent_width
- comma_first
- ...

So, in sqlparse you can influence these things, but blackbricks "defines" SQL formatting standards, which can not be changed/influenced.

In my previous projects we agreed on "project-own formatting defaults", put them into a config filed, shared this within the project, and the code was well formatted. And there were no changes caused by formatting options. For this we could focus on "real" changes in the code when reviewing PRs.

**Comment by bsamseth** [comment]
Apologies for not getting back on this in a reasonable time.

I thought about this some more, and I'm not convinced my idea of trying to perfectly match Databricks's SQL formatting is a worthwhile effort. Instead, in the detail of #40 I noticed that sqlparse now apparently reads configuration from `pyproject.toml`. That means that instead of mirroring all options of both black and sqlparse in blackbricks, it could instead let them configure themselves natively, and blackbricks could respect those configs. Done right that would remove the need to set configs via blackbricks, and support any and all existing and future configs that might be added to either one. In particular to this issue, all the options would become available via sqlparse's native config. 

I've been wanting to do this for black anyway, and now that sqlparse allows similar functionality I think it might be time. That'll be a separate issue, see below 👇🏻

**Comment by bsamseth** [comment]
Nvm, I spoke too soon. sqlparse does not allow config via pyproject.toml, instead they just switched to using it within their repo.
